File: aula-02/src/Board.java
class Board {
    private int[][] state;
    private int currentX; // current position (row)
    private int currentY; // current position (column)
    private final int boardSize; // Board dimension

    public Board(int[][] initialState, int boardSize) {
        this.state = initialState;
        this.boardSize = boardSize;
        // Find the initial position (value 0)
        findInitialPosition();
    }

    private void findInitialPosition() {
        for (int i = 0; i < boardSize; i++) {
            for (int j = 0; j < boardSize; j++) {
                if (state[i][j] == 0) {
                    currentX = i;
                    currentY = j;
                    return;
                }
            }
        }
    }

    public void executeMoves(String moves) {
        for (char move : moves.toCharArray()) {
            switch (move) {
                case 'u': moveUp(); break;
                case 'd': moveDown(); break;
                case 'l': moveLeft(); break;
                case 'r': moveRight(); break;
            }
            displayState();
        }
    }

    private void moveUp() {
        if (currentX < boardSize - 1) {
            swapPositions(currentX, currentY, currentX + 1, currentY);
            currentX++;
        }
    }

    private void moveDown() {
        if (currentX > 0) {
            swapPositions(currentX, currentY, currentX - 1, currentY);
            currentX--;
        }
    }

    private void moveLeft() {
        if (currentY < boardSize - 1) {
            swapPositions(currentX, currentY, currentX, currentY + 1);
            currentY++;
        }
    }

    private void moveRight() {
        if (currentY > 0) {
            swapPositions(currentX, currentY, currentX, currentY - 1);
            currentY--;
        }
    }

    private void swapPositions(int x1, int y1, int x2, int y2) {
        int temp = state[x1][y1];
        state[x1][y1] = state[x2][y2];
        state[x2][y2] = temp;
    }

    public void displayState() {
        // Print the game board state in a formatted grid
        for (int i = 0; i < boardSize; i++) {
            // Print top border of the row
            System.out.print("+");
            for(int k = 0; k < boardSize; k++){
                System.out.print("------+");  // Each cell has a 7-character wide border
            }
            
            System.out.print("\n");  // New line after the border
            
            // Print cell values row
            for (int j = 0; j < boardSize; j++) {
                // Format each number to take exactly 2 spaces (right-aligned)
                // %2d means:
                // - %d for integer
                // - 2 for minimum width of 2 characters
                System.out.printf("|  %2d  ", state[i][j]);
            }
            System.out.print("|");  // Close the last cell in the row
            System.out.println();   // New line after values
        }
        
        // Print bottom border of the board
        System.out.print("+");
        for(int k = 0; k < boardSize; k++){
            System.out.print("------+");
        }
        System.out.println("\n");  // Extra new line for spacing
    }

    public boolean isWin() {
        int aux = 0;
        for (int i = 0; i < boardSize; i++) {
            for (int j = 0; j < boardSize; j++) {
                if(aux != state[i][j]) return false;
                aux++;
            }
        }
        return true;
    }
}

File: aula-03/Board.py
class Board:
    def __init__(self, initial_state, board_size):
        # Create a deep copy of the initial state to avoid modifying the original
        self.state = [row[:] for row in initial_state]  
        self.board_size = board_size
        # Find and set the initial position of the empty tile (0)
        self.current_x, self.current_y = self.find_initial_position()

    def find_initial_position(self):
        # Search the board to find the position of the empty tile (0)
        for i in range(self.board_size):
            for j in range(self.board_size):
                if self.state[i][j] == 0:
                    return i, j
        return 0, 0  # Default position if not found (shouldn't happen in valid puzzles)

    def execute_moves(self, moves):
        # Execute a sequence of moves (u=up, d=down, l=left, r=right)
        for move in moves:
            if move == 'u':
                self.move_up()
            elif move == 'd':
                self.move_down()
            elif move == 'l':
                self.move_left()
            elif move == 'r':
                self.move_right()
            # Display the board state after each move
            self.display_state()

    def move_up(self):
        # Move the empty tile up by swapping with the tile below
        if self.current_x < self.board_size - 1:
            self.swap_positions(self.current_x, self.current_y, self.current_x + 1, self.current_y)
            self.current_x += 1

    def move_down(self):
        # Move the empty tile down by swapping with the tile above
        if self.current_x > 0:
            self.swap_positions(self.current_x, self.current_y, self.current_x - 1, self.current_y)
            self.current_x -= 1

    def move_left(self):
        # Move the empty tile left by swapping with the tile to the right
        if self.current_y < self.board_size - 1:
            self.swap_positions(self.current_x, self.current_y, self.current_x, self.current_y + 1)
            self.current_y += 1

    def move_right(self):
        # Move the empty tile right by swapping with the tile to the left
        if self.current_y > 0:
            self.swap_positions(self.current_x, self.current_y, self.current_x, self.current_y - 1)
            self.current_y -= 1

    def swap_positions(self, x1, y1, x2, y2):
        # Swap the values of two positions on the board
        self.state[x1][y1], self.state[x2][y2] = self.state[x2][y2], self.state[x1][y1]

    def display_state(self):
        # Print the game board in a formatted grid
        for i in range(self.board_size):
            # Print top border of the row
            print("+", end="")
            for k in range(self.board_size):
                print("------+", end="")  # Each cell has a 7-character wide border
            print()  # New line after the border
            
           # Print cell values row
            for j in range(self.board_size):
                value = self.state[i][j]
                # If the value is 0, print a space; otherwise, print the number
                print(f"|  {'  ' if value == 0 else f'{value:2d}'}  ", end="")
            print("|")  # Close the last cell in the row
        
        # Print bottom border of the board
        print("+", end="")
        for k in range(self.board_size):
            print("------+", end="")
        print("\n")  # Extra new line for spacing

    def is_win(self):
        # Check if the current state is the winning configuration (sequential numbers)
        aux = 0
        for i in range(self.board_size):
            for j in range(self.board_size):
                if aux != self.state[i][j]:
                    return False
                aux += 1
        return True